fortitude/src/rules/correctness/nonportable_shortcircuit_inquiry.rs

```diff
@@ -0,0 +1,182 @@
+///   integer, optional, intent(in) :: arg1
+///
+///   if (present(arg1) .and. arg1 > 2) then
+///     example = arg1 * arg1
+///   else
+///     example = 1
+///   end if
+/// ```
+/// The compiler may or may not evaluate `arg1 > 2` _even if_ `present(arg1)` is
+/// false. This is a runtime error, and may crash your program.
+///
+/// Use instead, noting that we either need to duplicate the `else` branch, or refactor
+/// differently:
+/// ```f90
+/// integer function example(arg1)
+///   integer, optional, intent(in) :: arg1
+///
+///   if (present(arg1)) then
+///     if (arg1 > 2) then
+///       example = arg1 * arg1
+///     else
+///       example = 1
+///     end if
+///   else
+///     example = 1
+///   end if
+/// ```
+///
+/// Or with Fortran 2023 (not currently supported by most compilers!):
+/// ```f90
+/// integer function example(arg1)
+///   integer, optional, intent(in) :: arg1
+///
+///   example = present(arg1) ? (arg1 > 2 ? arg1 * arg1 : 1) : 1
+/// ```
+/// Note that although the true/false arms of the conditional-expression are lazily
+/// evaluated, it's still not possible to use a compound logical expression here, so we
+/// still must have a nested expression and duplicate the default value.
+///
+/// ## References
+/// - <https://www.scivision.dev/fortran-short-circuit-logic/>
+#[derive(ViolationMetadata)]
+pub(crate) struct NonportableShortcircuitInquiry {
+    arg: String,
+    function: String,
+    // Useful for when we get multiple notes in DiagnosticMessages
+    #[allow(dead_code)]
+    present: TextRange,
+}
+
+impl Violation for NonportableShortcircuitInquiry {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { arg, function, .. } = self;
+        format!("variable inquiry `{function}({arg})` and use in same logical expression")
+    }
+}
+
+#[derive(Debug)]
+struct PresentCall {
+    pub range: TextRange,
+    pub arg: String,
+}
+
+fn present_call(expr: &Node, src: &str, function: &str) -> Option<PresentCall> {
+    if expr.kind() != "call_expression" {
+        return None;
+    }
+    if expr.child(0)?.to_text(src)?.to_lowercase() != function {
+        return None;
+    }
+    let arg_list = expr.child_with_name("argument_list")?;
+    // Make sure we skip the two-arg version of `associated`
+    // Length 3: "(", "identifier, ")"
+    if arg_list.children(&mut arg_list.walk()).len() != 3 {
+        return None;
+    }
+
+    let identifier = arg_list.child_with_name("identifier")?;
+    let arg = identifier.to_text(src)?.to_lowercase().to_string();
+
+    Some(PresentCall {
+        range: expr.textrange(),
+        arg,
+    })
+}
+
+impl AstRule for NonportableShortcircuitInquiry {
+    fn check<'a>(
+        _settings: &Settings,
+        node: &'a Node,
+        src: &'a SourceFile,
+    ) -> Option<Vec<Diagnostic>> {
+        let expr = node.child(1)?;
+        let text = src.source_text();
+
+        let violations = ["present", "allocatable", "associated"]
```

Comment:
```suggestion
        let violations = ["present", "allocated", "associated"]
```

Reply:
Actually they were, I only noticed because an `allocated` in the tests wasn't triggering it!